===================================================================
RCS file: /cvsroot/mozilla/content/base/src/ nsGenericElement.cpp,v
retrieving revision 3.595
retrieving revision 3.596
Index: mozilla/content/base/src/nsGenericElement.cpp
===================================================================
--- mozilla.orig/content/base/src/nsGenericElement.cpp
+++ mozilla/content/base/src/nsGenericElement.cpp
@@ -78,16 +78,17 @@
 #include "nsIScriptSecurityManager.h"
 #include "nsIDOMMutationEvent.h"
 #include "nsMutationEvent.h"
 #include "nsNodeUtils.h"
 #include "nsDocument.h"
 #ifdef MOZ_XUL
 #include "nsXULElement.h"
 #endif
+#include "nsFrameManager.h"
 
 #include "nsBindingManager.h"
 #include "nsXBLBinding.h"
 #include "nsIDOMCSSStyleDeclaration.h"
 #include "nsIDOMViewCSS.h"
 #include "nsIXBLService.h"
 #include "nsPIDOMWindow.h"
 #include "nsIBoxObject.h"
@@ -651,29 +652,24 @@
                                            nsIDOMNodeList** aReturn)
 {
   return nsDocument::GetElementsByClassNameHelper(mContent, aClasses, aReturn);
 }
 
 static nsPoint
 GetOffsetFromInitialContainingBlock(nsIFrame* aFrame)
 {
-  nsPresContext* presContext = aFrame->PresContext();
-  nsIPresShell* shell = presContext->PresShell();
-  nsIFrame* rootScrollFrame = shell->GetRootScrollFrame();
+  nsIFrame* rootFrame = aFrame->PresContext()->FrameManager()->GetRootFrame();
   nsPoint pt(0,0);
-  nsIFrame* child = aFrame;
-  for (nsIFrame* p = aFrame->GetParent(); p && p != rootScrollFrame;
-       p = p->GetParent()) {
-    pt += p->GetPositionOfChildIgnoringScrolling(child);
+  for (nsIFrame* p = aFrame; p != rootFrame; p = p->GetParent()) {
     // coordinates of elements inside a foreignobject are relative to the top-left
     // of the nearest foreignobject
-    if (p->IsFrameOfType(nsIFrame::eSVGForeignObject))
+    if (p->IsFrameOfType(nsIFrame::eSVGForeignObject) && p != aFrame)
       return pt;
-    child = p;
+    pt += p->GetPosition();
   }
   return pt;
 }
 
 static double
 RoundFloat(double aValue)
 {
   return floor(aValue + 0.5);
